aufs: apply dirperm1 by default if supported

dqminh opened the pull request:
Automatically detect support for aufs `dirperm1` option and apply it.
`dirperm1` tells aufs to check the permission bits of the directory on the
topmost branch and ignore the permission bits on all lower branches.
It can be used to fix aufs' permission bug (i.e., upper layer having
broader mask than the lower layer).

`dirperm1` man page is at: http://aufs.sourceforge.net/aufs3/man.html

Fixes #783

dqminh committed:
aufs: apply dirperm1 by default if supported

Automatically detect support for aufs `dirperm1` option and apply it.
`dirperm1` tells aufs to check the permission bits of the directory on the
topmost branch and ignore the permission bits on all lower branches.
It can be used to fix aufs' permission bug (i.e., upper layer having
broader mask than the lower layer).

More information about the bug can be found at https://github.com/docker/docker/issues/783
`dirperm1` man page is at: http://aufs.sourceforge.net/aufs3/man.html

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh commented:
cc @tiborvass 

Also it seems like enabling `dirperm1` makes remounting layer significantly slower (1+ms in my VM) ( happens when the user has a lot of layers that total length of mount options exceeds PAGE_SIZE ). `TestMountMoreThan42Layers` is much more slower with this patch.

tiborvass commented:
+1 for this, I like it.

Ping @vbatts @unclejack feel free to step in to say if you don't like the way this is implemented.

tiborvass commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -460,3 +472,36 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
+			err         error
+		)
+
+		if base, err = ioutil.TempDir("", "docker-aufs-base"); err != nil {
```
should be 
```Go
base, err := ioutil.TempDir("", "docker-aufs-base")
if err != nil {
```

Same for `union, err :=`

tiborvass commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -460,3 +472,36 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
+		defer os.RemoveAll(union)
+
+		opts := fmt.Sprintf("br:%s,dirperm1,xino=/dev/shm/aufs.xino", base)
+		if err = mount("none", union, "aufs", 0, opts); err != nil {
```
`err :=`

tiborvass commented:
@jfrazelle this would be a great fix that a lot of people would welcome. Also it would close an issue with a 3-digit number :P

@dqminh would you mind removing the part that mentions the issue in some KNOWN issues section somewhere?

dqminh commented:
> would you mind removing the part that mentions the issue in some KNOWN issues section somewhere?

@tiborvass this won't fix it for system that doesnt have `dirperm1` patch though, so I think we still need to keep the note. I can add some note about `dirperm1` being set as the default option if it's available.

tiborvass commented:
@dqminh fair enough! Thanks :)

dqminh committed:
document dirperm1 fix for #783 in known issues

Since `dirperm1` requires a more recent aufs patch than many current OS release,
we cant remove #783 completely. This documents that docker will apply `dirperm1`
automatically for systems that support it

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

crosbymichael commented:
LGTM

dqminh committed:
print dirperm1 supported status in docker info

It's easier for users to check if their systems support dirperm1 just by using
docker info

Signed-off-by: Daniel, Dao Quang Minh <dqminh89@gmail.com>

dqminh commented:
@tiborvass i added dirperm1 support status to `docker info` output too so there's an easy way to tell user whether their system is still affected by the permission bug.

dqminh commented:
@tiborvass is this merge-able for 1.6, or is it too late now ?

jfrazelle commented:
I added the milestone @crosbymichael is that ok with you

vbatts commented:
LGTM

Blaisorblade commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -460,3 +473,33 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
+		defer os.RemoveAll(union)
+
+		opts := fmt.Sprintf("br:%s,dirperm1,xino=/dev/shm/aufs.xino", base)
+		if err := mount("none", union, "aufs", 0, opts); err != nil {
```
While investigating some other (transient) docker problem, I saw this error (once) in my kernel log — is it a harmless error produced here while testing the presence of `dirperm1`? Google wasn't very helpful.
```
[   16.825747] aufs au_opts_parse:1155:docker[1679]: unknown option dirperm1
```

Blaisorblade replied:
OK, I'm guessing yes:
```
# docker info | grep Dirperm1
 Dirperm1 Supported: false
```
